### ISSUE TITLE
Cannot use generic parameters of a constant's containing function

### ISSUE THREAD

**Issue by chrisduerr** [rust-lang/rust#54327]
So I'm not quite certain what exactly causes this issue, but I've noticed a compiler panic when using `static_assertions` with a generic `Box<T>`.

Here's the code snippet which panics:
(_**NOTE(@eddyb)**: I've added the original definition, as the crate has changed_)
```rust
// #[macro_use]
// extern crate static_assertions;

macro_rules! const_assert {
    ($cond:expr) => {
        // Causes overflow if condition is false
        let _ = [(); 0 - (!($cond) as usize)];
    };
}

fn main() {}

fn test<T>() {
    const_assert!(::std::mem::size_of::<Box<T>>());
}
```

This will fail with the latest version of the `static_assertions` crate. This is the compiler error which is reported when this code is run:
```
   Compiling testing v0.1.0 (file:///tmp/testing)
error: internal compiler error: librustc/ty/subst.rs:489: Type parameter `T/#0` (T/0) out of range when substituting (root type=Some(fn() -> usize {std::mem::size_of::<std::boxed::Box<T>>})) substs=[]

thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:524:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.30.0-nightly (f2302daef 2018-09-12) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `testing`.

To learn more, run the command again with --verbose.
```

**Comment by csmoe** [comment]
Build error from `rustc-1.20.0`:
```rust
error[E0015]: calls in constants are limited to struct and enum constructors
 --> src/main.rs:7:19
  |
7 |     const_assert!(::std::mem::size_of::<Box<T>>());
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: a limited form of compile-time function evaluation is available on a nightly compiler via `const fn`
 --> src/main.rs:7:19
  |
7 |     const_assert!(::std::mem::size_of::<Box<T>>());
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

**Comment by chrisduerr** [comment]
Oh yeah, I forgot to mention that I tested both stable and nightly. Thanks for testing it yourself.

**Comment by pnkfelix** [comment]
visited for triage. P-high.

**Comment by nikomatsakis** [comment]
Can we bisect this using https://github.com/rust-lang-nursery/cargo-bisect-rustc? It seems like an old regression, but we could at least get down to a nightly build.

**Comment by eddyb** [comment]
Does `const_assert!` use an array length to trigger this? If so, it shouldn't ever worked, since we never supported that, so I don't know how it started compiling *in the first place*.

**Comment by eddyb** [comment]
@nikomatsakis for the record, I don't think this needs bisected. I just checked and `size_of` was `const fn`-unstable in 1.23.0 and became `const fn`-stable in 1.24.0.
Can we remove the regression tag? The ICE is just #43408 AFAICT.

**Comment by pnkfelix** [comment]
visited for triage during T-compiler weekly meeting.

Team basically agreed that:
 * This should not be tagged as a regression, under reasoning outlined by @eddyb above
 * This is sort of a duplicate of #43408, except that issue *might* be best off dedicated to correctly implementing the feature (of e.g. using `size_of` in generic const array length contexts)
 * while we may want a separate issue to track effort to do a controlled error exit in this scenario (rather than an ICE) in the meantime.

**Comment by pnkfelix** [comment]
visited for triage during T-compiler weekly meeting.

This probably doesn't need to be P-high. The remaining issue is about implementing a controlled error exit with a nicer diagnostic.

But it has been claimed (I think by @eddyb ) that we cannot implement that without lazy normalization.

Assigning to @nikomatsakis and @eddyb to resolve the questions here.

**Comment by pnkfelix** [comment]
visited for triage. Downgrading to P-medium. Tagging as S-blocked (on lazy normalization).

**Comment by jakubadamw** [comment]
With the latest nightly this no longer ICEs and instead produces the following errors:

```
error[E0401]: can't use generic parameters from outer function
 --> src\lib.rs:7:45
  |
6 | fn test<T>() {
  |    ---- - type parameter from outer function
  |    |
  |    try adding a local generic parameter in this method instead
7 |     const_assert!(::std::mem::size_of::<Box<T>>());
  |                                             ^ use of generic parameter from outer function

error[E0308]: mismatched types
 --> src\lib.rs:7:19
  |
7 |     const_assert!(::std::mem::size_of::<Box<T>>());
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected bool, found usize

error[E0080]: evaluation of constant value failed
 --> src\lib.rs:7:5
  |
7 |     const_assert!(::std::mem::size_of::<Box<T>>());
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ referenced constant has errors
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0080, E0308, E0401.
For more information about an error, try `rustc --explain E0080`.
```

**Comment by eddyb** [comment]
@jakubadamw Is that using the same version of `const_assert`?
You should expand it to see what it actually does: if it now creates a nested `const` item, that can't refer to type parameters *anyway*, regardless of #43408.

**Comment by jakubadamw** [comment]
@eddyb, I don't know what version was used originally but I'm using the newest one.
I should have been more clear. I'm not surprised this errors out. In fact, what I meant to emphasize is just that it doesn't ICE any more so we can take off the label.

Here's the output of the expansion:
```rust
#![feature(prelude_import)]
#![no_std]
#[prelude_import]
use ::std::prelude::v1::*;
#[macro_use]
extern crate std;
#[macro_use]
extern crate static_assertions;

fn main() { }

fn test<T>() {
    #[allow(unknown_lints, eq_op)]
    let _ =
        [();
            0 -
                !({
                      const B: bool = ::std::mem::size_of::<Box<T>>();
                      B
                  }) as usize];
```

**Comment by eddyb** [comment]
Yeah, it changed:
* https://docs.rs/static_assertions/0.1.0/src/static_assertions/lib.rs.html#90-101
* https://docs.rs/static_assertions/0.3.3/src/static_assertions/const_assert.rs.html#85

I've just updated the top comment to a version which still ICEs (which makes sense, 0 progress has been made in that area).

Closing as duplicate of #43408.
